File: apps/mobile/lib/stripe.ts
// Client-side Stripe tokenization for payout bank accounts.
// Uses the publishable key to send raw bank details directly to Stripe's
// token API — the account number never touches the Tarea backend. The
// resulting token is exchanged for an external account server-side.

const PUBLISHABLE_KEY = process.env.EXPO_PUBLIC_STRIPE_PUBLISHABLE_KEY ?? "";

type BankTokenResult = { id?: string; error?: string };

export async function createBankAccountToken(params: {
  routingNumber: string;
  accountNumber: string;
  accountHolderName: string;
}): Promise<BankTokenResult> {
  if (!PUBLISHABLE_KEY || PUBLISHABLE_KEY.includes("placeholder")) {
    return { error: "Payments are not configured. Please contact support." };
  }

  const body = new URLSearchParams({
    "bank_account[country]": "US",
    "bank_account[currency]": "usd",
    "bank_account[routing_number]": params.routingNumber,
    "bank_account[account_number]": params.accountNumber,
    "bank_account[account_holder_name]": params.accountHolderName,
    "bank_account[account_holder_type]": "individual",
  });

  try {
    const res = await fetch("https://api.stripe.com/v1/tokens", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${PUBLISHABLE_KEY}`,
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: body.toString(),
    });
    const data = await res.json();
    if (!res.ok) {
      return { error: data?.error?.message ?? "Could not validate your bank details." };
    }
    return { id: data.id as string };
  } catch {
    return { error: "Network error. Please check your connection and try again." };
  }
}

type CardTokenResult = { id?: string; error?: string };

// Tokenize a debit card for instant payouts. Card number/CVC go directly to
// Stripe — never to the Tarea backend. `currency` makes it a payout-capable
// external-account token (required to attach it to the connected account).
export async function createCardToken(params: {
  number: string;
  expMonth: string;
  expYear: string;
  cvc: string;
  name?: string;
}): Promise<CardTokenResult> {
  if (!PUBLISHABLE_KEY || PUBLISHABLE_KEY.includes("placeholder")) {
    return { error: "Payments are not configured. Please contact support." };
  }

  const body = new URLSearchParams({
    "card[number]": params.number,
    "card[exp_month]": params.expMonth,
    "card[exp_year]": params.expYear,
    "card[cvc]": params.cvc,
    "card[currency]": "usd",
    ...(params.name ? { "card[name]": params.name } : {}),
  });

  try {
    const res = await fetch("https://api.stripe.com/v1/tokens", {
      method: "POST",
      headers: {
        Authorization: `Bearer ${PUBLISHABLE_KEY}`,
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: body.toString(),
    });
    const data = await res.json();
    if (!res.ok) {
      return { error: data?.error?.message ?? "Could not validate your card details." };
    }
    if (data?.card?.funding && data.card.funding !== "debit") {
      return { error: "That looks like a credit card. Instant payouts require a debit card." };
    }
    return { id: data.id as string };
  } catch {
    return { error: "Network error. Please check your connection and try again." };
  }
}

File: apps/mobile-flutter/lib/stripe.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

// Client-side Stripe tokenization for payout methods. Raw bank/card details go
// straight to Stripe with the PUBLISHABLE key (safe to embed) — they never
// touch the Tarea backend. The returned token is exchanged for an external
// account server-side via /handyman/payout-methods.
const String _publishableKey =
    'pk_live_51TWf7zGua4dfcnJ9ZIIVfje6IJGveoQZthhDr54jxVYLwGM5xTbdhAC4hY2Z0SiOHXPRJwmAPOpFXom7YZIRsXZd000tuovwCs';

class StripeTokenResult {
  final String? id;
  final String? error;
  const StripeTokenResult({this.id, this.error});
}

class StripeTokens {
  static Future<StripeTokenResult> bankAccount({
    required String routingNumber,
    required String accountNumber,
    required String accountHolderName,
  }) =>
      _token({
        'bank_account[country]': 'US',
        'bank_account[currency]': 'usd',
        'bank_account[routing_number]': routingNumber,
        'bank_account[account_number]': accountNumber,
        'bank_account[account_holder_name]': accountHolderName,
        'bank_account[account_holder_type]': 'individual',
      });

  static Future<StripeTokenResult> card({
    required String number,
    required String expMonth,
    required String expYear,
    required String cvc,
    String name = '',
  }) =>
      _token({
        'card[number]': number,
        'card[exp_month]': expMonth,
        'card[exp_year]': expYear,
        'card[cvc]': cvc,
        'card[currency]': 'usd',
        if (name.isNotEmpty) 'card[name]': name,
      });

  static Future<StripeTokenResult> _token(Map<String, String> body) async {
    if (_publishableKey.isEmpty || _publishableKey.contains('placeholder')) {
      return const StripeTokenResult(error: 'Payments are not configured.');
    }
    try {
      final res = await http.post(
        Uri.parse('https://api.stripe.com/v1/tokens'),
        headers: {
          'Authorization': 'Bearer $_publishableKey',
          'Content-Type': 'application/x-www-form-urlencoded',
        },
        body: body,
      );
      final data = jsonDecode(res.body);
      if (res.statusCode >= 200 && res.statusCode < 300) {
        return StripeTokenResult(id: (data['id'] ?? '').toString());
      }
      return StripeTokenResult(error: (data['error']?['message'] ?? 'Could not validate your details.').toString());
    } catch (_) {
      return const StripeTokenResult(error: 'Network error. Please try again.');
    }
  }
}
